operator/v1/types_network.go

```diff
@@ -388,6 +392,31 @@ type HybridOverlayConfig struct {
 type IPsecConfig struct {
 }
 
+// GatewayMode holds the node gateway mode
+type GatewayMode string
+
+const (
+	// GatewayModeShared indicates OVN shares a gateway interface with the node
+	GatewayModeShared GatewayMode = "shared"
+	// GatewayModeLocal indicates OVN creates a local NAT-ed interface for the gateway
+	GatewayModeLocal GatewayMode = "local"
+)
+
+// GatewayConfig holds node gateway-related parsed config file parameters and command-line overrides
+type GatewayConfig struct {
+	// mode is the gateway mode; if may be either "shared", or "local".
+	// Default is shared.
+	// +kubebuilder:validation:Enum="shared";"local"
+	// +kubebuilder:default:="shared"
+	// +optional
+	Mode GatewayMode `json:"mode,omitempty"`
+	// disableSNATMultipleGws sets whether to disable SNAT of egress traffic in namespaces annotated with routing-external-gws.
+	// Default is false.
+	// +kubebuilder:default:=false
+	// +optional
+	DisableSNATMultipleGWs bool `json:"disableSNATMultipleGWs,omitempty"`
```

Comment:
Isn't this part of a feature that is not supported for most customers? If so, it's weird to put it in the API...

Reply:
> The CNO only allows toggling this because it breaks egress IP completely, see: [openshift/cluster-network-operator#1145](https://github.com/openshift/cluster-network-operator/pull/1145) but I am not sure should be exposed to customers. And if we do: it should have a better name than `DisableSNATMultipleGWs` (which is completely incomprehensible), maybe `EnableICNI` would be clearer (because it's only used within the scope of ICNI and its F5 integration)?

yeah you make a very good point, this option right now like you said is needed only for ICNI model. So if we feel this isn't something we should expose and should just continue using the `gateway-mode-config` ConfigMap, I'm fine with that. Also we don't have any documentation on how to use this feature, so maybe the hidden feature doesn't need the hidden flag exposed :)

Reply:
I don't object to the idea of having a public "enable ICNI" flag (with documentation explaining that it's an unsupported alpha feature), but if we're going to put it in the API, we need to decide whether we are really going to call it "ICNI" forever. And also, should the flag distinguish ICNI 1 vs 2?

Reply:
> I personally am against exposing all of this FWIW. Exposing these knobs to customers means we will need to support shared and local gateway mode for OVN-Kubernetes for ever. I for one was hoping that whatever networking feature/capability was provided by local but not shared, would be implemented in the future and covered by some epic/user story. From my understanding: the sole argument for using local gateway mode today is: you, as a user, want to modify the kernel routing table or you manage multiple NICs/routes yourself, and shared gateway mode currently breaks you because we haven't implemented what they are doing themselves in OVN-Kubernetes (or an API for them to do that). I might be missing something critical though, which makes it so that we can't align and fill these needs?
> 

Yes, I think we all intended to make LGW go away at some point and only have SGW (mainly for HW offloading as I understand). Unfortunately a lot of users actually count on having the egress traffic coming from the pod leave via the management port into the kernel before sending it out from the host. This is simply never going to happen with SGW because all traffic from pods will leave via br-ex and won't touch the host stack.

Implementing this in SGW would mean adding(hacking) necessary routes in `ovn_cluster_router` to ask it to send it out into the host. The plan currently is to provide like a CRD or something for SGW to allow users to be able to specify routes. Trouble is this means people need to know the OVN topology :/ .

Anyways until a proper solution in SGW gets shaped up, the other solution was to support LGW since we can't really regress (but I guess that ship has sailed - lucky saved by the hidden mode in the config-map).

The middle ground solution is to have a toggle in SGW where we allow users to say 1) I want pod egress to go via host 2) I am ok with it not touching host. This former toggle will now be called "local gateway". At least that's the aim if I can get the "move lgw closer to sgw" epic done properly. If we can remove all the extra bits from lgw, then its just a toggle for how we wanna direct traffic.

> EDIT: well I am more so referring to the gateway mode flag. ICNI and egress IP are logically incompatible, so customers need to chose one or the other. But the impact of that flag on OVN-Kubernetes as a project is much smaller than supporting both gateway modes.

right.

Reply:
DisableSNATMultipleGWs isn't to enable/disable ICNI. ICNI is always enabled. The flag simply decides whether ICNI egress traffic will be SNAT'ed to the hosts IP or expose the pod source IPs directly. For our current use case we disable SNAT and expose pod IPs. We have never purposefully exposed this so that someone could use ICNI with SNAT. I would say we can hold off on exposing it for now, but keep the gatewayconfig block so we could add it (or other features later).

Reply:
> DisableSNATMultipleGWs isn't to enable/disable ICNI. ICNI is always enabled. The flag simply decides whether ICNI egress traffic will be SNAT'ed to the hosts IP or expose the pod source IPs directly. For our current use case we disable SNAT and expose pod IPs. We have never purposefully exposed this so that someone could use ICNI with SNAT. I would say we can hold off on exposing it for now, but keep the gatewayconfig block so we could add it (or other features later).

Oh okay, this is news to me :D I always thought the disableSNAT's option was also needed for ICNI to work, but it seems like we first hit the ecmp route and only then SNAT towards the end, so makes sense that SNAT is not a strict requirement for ICNI to work, just that outgoing traffic won't have the podIP's as their source. Should we just add the toggle in this PR itself? Asking because in 4.9 we were actually allowing the user to specify this via the hidden config-map. Now from 4.10 if we suddenly remove that config-map and don't provide a way to specify this option are we in some way regressing?

Reply:
I think the purpose of it in 4.9 was to workaround the incompatibility with egress IP right? In 4.10 we are going to make egress gw and egress IP compatible so that isn't a valid reason to expose the option. If we officially supported multiple exgw then I would agree we should keep the option and expose it via API, but we do not officially support it for 4.10. Therefore I think we should hold off on exposing it until we officially support it, and ensure that multiple exgw and egress IP are compatible in 4.10 (which you are already working  on).

Reply:
>  think the purpose of it in 4.9 was to workaround the incompatibility with egress IP right? In 4.10 we are going to make egress gw and egress IP compatible so that isn't a valid reason to expose the option. If we officially supported multiple exgw then I would agree we should keep the option and expose it via API, but we do not officially support it for 4.10. Therefore I think we should hold off on exposing it until we officially support it, and ensure that multiple exgw and egress IP are compatible in 4.10 (which you are already working on).

yes, true, the intention was to allow exclusiveness and that has been removed with https://github.com/ovn-org/ovn-kubernetes/pull/2686.

Cool as long as we agree we have not agreed to support this feature through API until its fully Alpha, I'm good with this decision.